File: 0439-ternary-expression-parser/0439-ternary-expression-parser.py
class Solution:
    def isValidAtomic(self, s: str) -> bool:
        return s[0] in 'TF' and s[1] == '?' and s[2] in 'TF0123456789'\
    and s[3] == ':' and s[4] in 'TF0123456789'
    
    def solveAtomic(self, s: str) -> str:
        return s[2] if s[0] == 'T' else s[4]
    
    def parseTernary(self, expression: str) -> str:
        while len(expression) != 1:
            j = len(expression) - 1
            while not self.isValidAtomic(expression[j - 4 : j + 1]):
                j -= 1
            expression = expression[:j - 4] +\
            self.solveAtomic(expression[j - 4: j + 1]) + expression[j + 1:]
            
        return expression

File: 0439-ternary-expression-parser/0439-ternary-expression-parser.cpp
class Solution {
public:
    string parseTernary(string expression) {
        auto isValidAtomic = [](string s){
            return (s[0] == 'T' || s[0] == 'F') && s[1] == '?' && \
                ((s[2] >= '0' && s[2] <= '9') || s[2] == 'T' || s[2] == 'F') && s[3] == ':'\
                && ((s[4] >= 0 && s[4] <= '9') || s[4] == 'T' || s[4] == 'F');
        };
        
        auto solveAtomic = [](string s){
            return s[0] == 'T' ? s[2] : s[4];
        };
        
        while (expression.size() != 1){
            int j = expression.size() - 1;
            while (!isValidAtomic(expression.substr(j - 4, 5))) --j;
            expression = expression.substr(0, j - 4) + solveAtomic(expression.substr(j - 4, 5)) + expression.substr(j + 1);
        }
        
        return expression;
    }
};